### PR TITLE
add happyx

### PR DESCRIPTION
**Pull request by Ethosa**
Repo: https://github.com/HapticX/happyx

### CODE REVIEW

**Comment by waghanza** [comment]
Will include `happyx` for tomorrow.  does it mean it replace `akane` @Ethosa ?

**Comment by waghanza** [comment]
There is a compilation error on container
```nim
  Building server/hpx using c backend
oserr.nim(95)            raiseOSError
Error: unhandled exception: No such file or directory
Additional info: /usr/src/app/src [OSError]
```
The dockerfile is
```dockerfile
FROM nimlang/nim:1.6.12-alpine

RUN apk update && \
    apk upgrade --no-cache

  RUN apk --no-cache add pcre-dev

WORKDIR /usr/src/app

COPY . ./

ENV PATH $PATH:/root/.nimble/bin

RUN nimble install -y

  RUN nim c  --define:usePcreHeader --passL: "/usr/lib/libpcre.a"  \
    --excessiveStackTrace:off \
    -d:release \
    --passC:-flto \
    --passL:-flto \
    server.nim

FROM alpine

WORKDIR /usr/src/app

RUN apk update && \
    apk upgrade --no-cache

RUN apk --no-cache add openssl-dev pcre-dev

  COPY --from=0 /usr/src/app/server /usr/src/app/server

  CMD /usr/src/app/server
```

**Comment by Ethosa** [comment]
> Will include happyx for tomorrow. does it mean it replace akane @Ethosa ?

`akane` is legacy 👀

I can delete `akane` and add commit for this PR 🔨

**Comment by Ethosa** [comment]
> There is a compilation error on container

@waghanza 
How I can change dockerfile?

Need change
```dockerfile
RUN nimble install -y
```
to
```dockerfile
RUN nimble install -d -y
```

**Comment by waghanza** [comment]
By setting `build_command: nimble install -d -y` in `config.yaml`, and I have
```nim
Installing happyx@0.15.1
   Building happyx/hpx using c backend
/tmp/nimble_1/githubcom_HapticXhappyx_0.15.1/src/happyx/private/cmpltime.nim(2, 3) Warning: imported and not used: 'strformat' [UnusedImport]
/tmp/nimble_1/githubcom_HapticXhappyx_0.15.1/src/happyx/ssg/server.nim(105, 1) Warning: use exitprocs.addExitProc; addQuitProc is deprecated [Deprecated]
/tmp/nimble_1/githubcom_HapticXhappyx_0.15.1/src/hpx.nim(415, 17) template/generic instantiation of `createThread` from here
/nim/lib/system/threads.nim(358, 50) template/generic instantiation of `threadProcWrapper` from here
/nim/lib/system/threads.nim(197, 27) template/generic instantiation of `threadProcWrapStackFrame` from here
/nim/lib/system/threads.nim(188, 27) template/generic instantiation of `threadProcWrapDispatch` from here
/nim/lib/system/threads.nim(171, 5) Warning: The bare except clause is deprecated; use `except CatchableError:` instead [BareExcept]
/tmp/nimble_1/githubcom_HapticXhappyx_0.15.1/src/hpx.nim(429, 8) template/generic instantiation of `serve` from here
/tmp/nimble_1/githubcom_HapticXhappyx_0.15.1/src/happyx/ssg/server.nim(358, 27) template/generic instantiation of `async` from here
/nim/lib/pure/asyncmacro.nim(42, 5) Warning: The bare except clause is deprecated; use `except CatchableError:` instead [BareExcept]
   Success: happyx installed successfully.
      Info: Dependency on cligen@any version already satisfied
  Verifying dependencies for cligen@1.6.4
      Info: Dependency on regex@any version already satisfied
  Verifying dependencies for regex@0.20.2
      Info: Dependency on unicodedb@>= 0.7.2 already satisfied
  Verifying dependencies for unicodedb@0.12.0
      Info: Dependency on httpx@any version already satisfied
  Verifying dependencies for httpx@0.3.6
      Info: Dependency on ioselectors@>= 0.1.6 already satisfied
  Verifying dependencies for ioselectors@0.1.8
      Info: Dependency on wepoll@>= 0.1.0 already satisfied
  Verifying dependencies for wepoll@0.1.0
      Info: Dependency on illwill@any version already satisfied
  Verifying dependencies for illwill@0.3.0
      Info: Dependency on nimja@any version already satisfied
  Verifying dependencies for nimja@0.8.7
      Info: Dependency on websocketx@any version already satisfied
  Verifying dependencies for websocketx@0.1.2
      Info: Dependency on ws@>= 0.4.3 already satisfied
  Verifying dependencies for ws@0.5.0
      Info: Dependency on httpx@>= 0.1.8 already satisfied
  Verifying dependencies for httpx@0.3.6
      Info: Dependency on ioselectors@>= 0.1.6 already satisfied
  Verifying dependencies for ioselectors@0.1.8
      Info: Dependency on wepoll@>= 0.1.0 already satisfied
  Verifying dependencies for wepoll@0.1.0
 Installing server@0.15.1
   Building server/hpx using c backend
oserr.nim(95)            raiseOSError
Error: unhandled exception: No such file or directory
Additional info: /usr/src/app/src [OSError]
       Tip: 219 messages have been suppressed, use --verbose to show them.
     Error: Build failed for package: server
        ... Execution failed with exit code 1
        ... Command: /nim/bin/nim c --colors:on --noNimblePath -d:release -d:NimblePkgVersion=0.15.1 --path:/root/.nimble/pkgs/cligen-1.6.4 --path:/root/.nimble/pkgs/unicodedb-0.12.0 --path:/root/.nimble/pkgs/regex-0.20.2 --path:/root/.nimble/pkgs/wepoll-0.1.0 --path:/root/.nimble/pkgs/ioselectors-0.1.8 --path:/root/.nimble/pkgs/httpx-0.3.6 --path:/root/.nimble/pkgs/illwill-0.3.0 --path:/root/.nimble/pkgs/nimja-0.8.7 --path:/root/.nimble/pkgs/ws-0.5.0 --path:/root/.nimble/pkgs/httpx-0.3.6 --path:/root/.nimble/pkgs/ioselectors-0.1.8 --path:/root/.nimble/pkgs/wepoll-0.1.0 --path:/root/.nimble/pkgs/websocketx-0.1.2 --path:/root/.nimble/pkgs/happyx-0.15.1 --path:/root/.nimble/pkgs/cligen-1.6.4 --path:/root/.nimble/pkgs/regex-0.20.2 --path:/root/.nimble/pkgs/unicodedb-0.12.0 --path:/root/.nimble/pkgs/httpx-0.3.6 --path:/root/.nimble/pkgs/ioselectors-0.1.8 --path:/root/.nimble/pkgs/wepoll-0.1.0 --path:/root/.nimble/pkgs/illwill-0.3.0 --path:/root/.nimble/pkgs/nimja-0.8.7 --path:/root/.nimble/pkgs/websocketx-0.1.2 --path:/root/.nimble/pkgs/ws-0.5.0 --path:/root/.nimble/pkgs/httpx-0.3.6 --path:/root/.nimble/pkgs/ioselectors-0.1.8 --path:/root/.nimble/pkgs/wepoll-0.1.0 --hints:off -o:/usr/src/app/hpx /usr/src/app/src/hpx.nim
The command '/bin/sh -c nimble install -y' returned a non-zero code: 1
```

**Comment by waghanza** [comment]
sorry @Ethosa, mis-understood
1. Add `build_with: nimble install -d -y` to `conifg.yaml`
2. Replace `RUN nimble install -y` with `{{#build_with}}{{{.}}}{{/build_wth}}}` in `nim/Dockerfile`

**Comment by waghanza** [comment]
Seems to work with `-d`. What is this switch for ? @Ethosa

**Comment by Ethosa** [comment]
> What is this switch for ?

@waghanza 
This flag disables compilation and fetches only dependencies without compilation binaries (HappyX contains CLI binary file)

File: nim/Dockerfile [hunk 1]
```diff
@@ -13,7 +13,12 @@ COPY . ./
 
 ENV PATH $PATH:/root/.nimble/bin
 
-RUN nimble install -y
+{{#build_with}}
+  RUN {{{.}}}
+{{/build_wth}}}
+{{^build_with}}
+  RUN nimble install -y
+{{/build_wth}}
```
Comment (waghanza) [inline nim/Dockerfile:21]:
```suggestion
{{/build_with}}
```

File: nim/Dockerfile [hunk 2]
```diff
@@ -13,7 +13,12 @@ COPY . ./
 
 ENV PATH $PATH:/root/.nimble/bin
 
-RUN nimble install -y
+{{#build_with}}
+  RUN {{{.}}}
+{{/build_wth}}}
```
Comment (waghanza) [inline nim/Dockerfile:18]:
```suggestion
{{/build_with}}}
```

**Comment by Ethosa** [comment]
Also `-d` is shortcut for `--depsOnly`

**Review by waghanza — CHANGES_REQUESTED** [review]
I induced you into errors, sorry

**Comment by waghanza** [comment]
Useful for other `nim` **frameworks**, seems like interesting behavior for me facing production needs

File: nim/Dockerfile [hunk 3]
```diff
@@ -13,7 +13,12 @@ COPY . ./
 
 ENV PATH $PATH:/root/.nimble/bin
 
-RUN nimble install -y
+{{#build_with}}
+  RUN {{{.}}}
+{{/build_with}}}
+{{^build_with}}
+  RUN nimble install -y
+{{/build_with}}
```
Comment (waghanza) [inline nim/Dockerfile:21]:
```suggestion
{{#build_with}}
  RUN {{{.}}}
{{/build_with}}
{{^build_with}}
  RUN nimble install -y
{{/build_with}}
```

**Review by waghanza — CHANGES_REQUESTED** [review]

**Comment by waghanza** [comment]
Compilation is OK now, only implementation is not correct, some routes are not throwing expected result
https://github.com/the-benchmarker/web-frameworks/actions/runs/4972765750/jobs/8898179596?pr=6357#step:8:1

@Ethosa 

PS : I've pushed throw github UI

File: nim/happyx/config.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+framework:
+  github: HapticX/happyx
+  version: 0.15.1
```
Comment (waghanza) [inline nim/happyx/config.yaml:3]:
```suggestion
  version: 0.15
```

File: nim/happyx/server.nimble [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+author = "HapticX"
+version = "0.15.1"
+license = "GNU GPLv3"
+srcDir = "src"
+installExt = @["nim"]
+bin = @["hpx"]
+
+# Dependencies
+
+requires "happyx >= 0.15.1"
```
Comment (waghanza) [inline nim/happyx/server.nimble:13]:
```suggestion
requires "happyx >= 0.15 & < 0.16"
```

File: nim/happyx/server.nim [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import happyx
+
+
+serve("0.0.0.0", 3000):
+  get "/":
+    ""
+  
+  get "/user$id:int":
```
Comment (waghanza) [inline nim/happyx/server.nim:8]: I think

```suggestion
  get "/user/$id:int":
```

**Review by waghanza — CHANGES_REQUESTED** [review]

**Review by waghanza — APPROVED** [review]

**Comment by waghanza** [comment]
@Ethosa `happyx` has been included https://web-frameworks-benchmark.netlify.app/result?l=nim

**Pull request merged**